sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -0,0 +1,682 @@
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# Author: Christopher Moody <chrisemoody@gmail.com>
+# Author: Nick Travers <nickt@squareup.com>
+# Implementation by Chris Moody & Nick Travers
+# See http://homepage.tudelft.nl/19j49/t-SNE.html for reference
+# implementations and papers describing the technique
+
+
+from libc.stdlib cimport malloc, free
+from libc.stdio cimport printf
+from libc.math cimport sqrt
+cimport numpy as np
+import numpy as np
+
+cdef extern from "math.h":
+    float fabsf(float x) nogil
+
+# Round points differing by less than this amount
+# effectively ignoring differences near the 32bit 
+# floating point precision
+cdef float EPSILON = 1e-6
+
+# This is effectively an ifdef statement in Cython
+# It allows us to write printf debugging lines
+# and remove them at compile time
+cdef enum:
+    DEBUGFLAG = 0
+
+cdef extern from "time.h":
+    # Declare only what is necessary from `tm` structure.
+    ctypedef long clock_t
+    clock_t clock() nogil
+    double CLOCKS_PER_SEC
+
+
+cdef struct Node:
+    # Keep track of the center of mass
+    float[3] cum_com
+    # If this is a leaf, the position of the particle within this leaf 
+    float[3] cur_pos
+    # The number of particles including all 
+    # nodes below this one
+    long cum_size
+    # Number of particles at this node
+    long size
+    # Index of the particle at this node
+    long point_index
+    # level = 0 is the root node
+    # And each subdivision adds 1 to the level
+    long level
+    # Left edge of this node, normalized to [0,1]
+    float[3] le
+    # The center of this node, equal to le + w/2.0
+    float[3] c
+    # The width of this node -- used to calculate the opening
+    # angle. Equal to width = re - le
+    float[3] w
+
+    # Does this node have children?
+    # Default to leaf until we add particles
+    int is_leaf
+    # Keep pointers to the child nodes
+    Node *children[2][2][2]
+    # Keep a pointer to the parent
+    Node *parent
+    # Pointer to the tree this node belongs too
+    Tree* tree
+
+cdef struct Tree:
```

Comment:
I've now removed the dimension restriction (though I need to update the documentation)... so it can be `dimension == 2` or `dimension == 6` in principle. Should we stick to Node and Tree?